Adding a few Picard basecalling tasks.

nh13 opened the pull request.

nh13 commented:
@tfenne see 8e4ae73 for a travis-ci build issue fix.  Basically, the tests were using too much memory in the container environment (exit 137 from docker is a SIGKILL).

nh13 commented on tasks/src/main/scala/dagr/tasks/picard/IlluminaBasecallsToSam.scala:
```diff
@@ -0,0 +1,96 @@
+      .flatMap { cores =>
+        resources.subset(Cores(cores), Memory(cores * memoryPerThread.value))
+      }.headOption
+  }
```
I am not convinced this is the right strategy.  I have seen many cases where we have A LOT more memory per core than 1G and it still wont succeed.  I am tempted to implement something that will increase the memory per core, as well as reducing the # of cores.

nh13 replied:
Solved in an upcoming commit; discussed with you.

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/CollectIlluminaBasecallingMetrics.scala:
```diff
@@ -0,0 +1,56 @@
+import scala.collection.mutable.ListBuffer
+
+object CollectIlluminaBasecallingMetrics {
+  def toMetricsFileName(lane: Int): String = s".lane-$lane.metrics"
```
toMetricsExtension?

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/CollectIlluminaBasecallingMetrics.scala:
```diff
@@ -0,0 +1,56 @@
+                                        barcodesDir: Option[DirPath] = None
+                                       ) extends PicardTask with PicardMetricsTask {
+
+  override def in: FilePath = outputDir.getOrElse(basecallsDir).resolve("basecalling")
```
This is bizarre.  Why would we prefer the output directory as the input directory?

nh13 replied:
Good point, it _was_ to be able to set the output metrics file correctly, which is based on the input file name.  I could override the `PicardMetricsTask.pathPrefix` method, but instead I'll put a large comment here.

nh13 replied:
Actually, I think I can have a parameter `prefix: Option[Path]` and then it will make more sense.  I'll comment this well in the code.

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/CollectIlluminaLaneMetrics.scala:
```diff
@@ -0,0 +1,52 @@
+  * is provided with either, an error will be thrown.
+  */
+class CollectIlluminaLaneMetrics(runDirectory: DirPath,
+                                 basecallsDir: DirPath,
```
Shouldn't this just be called `output`?

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/ExtractIlluminaBarcodes.scala:
```diff
@@ -0,0 +1,76 @@
+                              compressOutputs: Boolean = true,
+                              minThreads: Int = 4,
+                              maxThreads: Int = 16,
+                              barcodesDir: Option[DirPath] = None
```
Again, I think this should be `output` because that's the name of the Picard parameter, and it really is just controlling where the output is going.

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/IlluminaBasecallsToSam.scala:
```diff
@@ -0,0 +1,96 @@
+                             includeNonPfReads: Boolean = false,
+                             minThreads: Int = 4,
+                             maxThreads: Int = 16,
+                             adapterPairs: Seq[IlluminaAdapterPair] = Seq(
```
This seems an odd choice.  I'd either pick a more sensible default set, or leave the tool to decide the defaults.

nh13 replied:
```
    @Option(doc = "Which adapters to look for in the read.")
    public List<IlluminaAdapterPair> ADAPTERS_TO_CHECK = new ArrayList<IlluminaAdapterPair>(
            Arrays.asList(IlluminaAdapterPair.INDEXED,
                    IlluminaAdapterPair.DUAL_INDEXED,
                    IlluminaAdapterPair.NEXTERA_V2,
                    IlluminaAdapterPair.FLUIDIGM));
```

https://github.com/broadinstitute/picard/blob/9f41e06267010c975a5e96913ab56a101f0f2b9f/src/main/java/picard/illumina/IlluminaBasecallsToSam.java#L198-L202

![](http://www.reactiongifs.com/wp-content/uploads/2013/10/tim-and-eric-mind-blown.gif)

nh13 commented:
@tfenne I fixed up the `CollectIlluminaBasecallingMetrics`, improved `IlluminaBasecallsToSam` based on discussion, and fixed all but the default adapters to check (which are the ones in Picard).  Back to you.

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/CollectIlluminaLaneMetrics.scala:
```diff
@@ -0,0 +1,52 @@
+
+/**
+  * CollectIlluminaLaneMetrics does not support read structures with molecular barcodes or skips.  If a read structure
+  * is provided with either, an error will be thrown.
```
What is the preferred solution?  To convert those operators to template?

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/ExtractIlluminaBarcodes.scala:
```diff
@@ -0,0 +1,76 @@
+    resources.subset(minCores=Cores(minThreads), maxCores=Cores(maxThreads), memory=Memory("4G"))
+  }
+
+  override def in: FilePath = _barcodesDir.resolve("barcode_counts")
```
Could you add a similar comment here please about overriding `in` to make `metricsFile` do the right thing?  Relatedly is this actually simpler than just in `addPicardArgs` writing:
 `buffer += "METRICS_FILE=" + _barcodesDir.resolve("barcode_counts") + metricsExtension`?

nh13 replied:
It's really to make it play nice with the `PicardMetricsTask`. I will update it to the same behavior.

tfenne commented on tasks/src/main/scala/dagr/tasks/picard/IlluminaBasecallsToSam.scala:
```diff
@@ -0,0 +1,96 @@
+  protected var memoryPerThread: Memory = Memory("2GB")
+
+  /** Increases the memory per core/thread and returns true if we can run with the fewest # of threads. */
+  override def retry(systemResources: SystemResources, taskInfo: TaskExecutionInfo): Boolean = {
```
I wonder if you should mix in one or more of the OOM traits so that this retry only happens if an OOM was detected and not some other kind of error?

nh13 commented:
@tfenne one last look before I merge?
